### ISSUE TITLE
automatick widget dialog open/show support for drag&drop/paste

### ISSUE THREAD

**Issue by voborl00** [ckeditor/ckeditor4#4359]
## Type of report

Feature request

When widget is constructed by its command (toolbar button) there is automatic dialog binding and its dialog windows pops up after widget is inserted into editor.

But when widgets is constructed on drag&dropped/pasted code fragment (and upcasted afterwards) there is no automatic dialog pop up, and it seems there is also no proper support for handling it manually. I only found this workaround based on editor's (after)Paste handler:

```
editor.on('paste', function(event) {
   var ckWidgetId = event.data.dataTransfer.getData('ckWidgetId');
   if(ckWidgetId && editor.widgets.registered[ckWidgetId]) {
      event.data.dataValue = editor.widgets.registered[ckWidgetId].template.source;
   }
});

... follows upcast(), init(), data() widget events

// workaround based on selected/focused widget
editor.on('afterPaste', function(event) {
	//editor.widgets.selected[0].edit();
	if(editor.widgets.focused) {
		editor.widgets.focused.edit();
	}
});
```

*What is the expected behavior of the proposed feature?*

Could you please point me to existing proper handling to be able to show widget's dialog upon drag&dropping/pasting it into editor or add some API support for handling it in widget command style (=automatic dialog binding). Thank you!

**Comment by voborl00** [comment]
I also found out another "hack", executing widget command from onPaste handler which opens widget's edit dialog

```
editor.on('paste', function(event) {
   var ckWidgetId = event.data.dataTransfer.getData('ckWidgetId');
   if(ckWidgetId && editor.widgets.registered[ckWidgetId]) {
      event.data.dataValue = '';
      editor.execCommand(ckWidgetId);
   }
});
```

but this only works with hard setting data.dataValue to empty string, any other not-empty string value (like widget's template property) is pasted into editor and even with "true" execCommand() boolean result outcome is based on (upcasted) widget template and not execCommand execution

**Comment by Dumluregn** [comment]
Hello, as far as I can see the code you provided:
```js
editor.on( 'afterPaste', function( event ) {
	if ( editor.widgets.focused ) {
		editor.widgets.focused.edit();
	}
} );
```
works - it triggers the dialog window after pasting and fills it with the correct data, and I think it's the right API for this purpose. Could you explain more thoroughly why aren't you satisfied with it?

**Comment by voborl00** [comment]
Hello, and thank you for reply. 

I find this approach depending on **paste** and **afterPaste** handlers too much complicated, as **afterPaste** reacts on any paste event so there has to be some "if" to tell whether it is the case with pasted widget, drag&dropped widget or pasted text, code, image, ...

Even if this "if" is acceptable, it's condition (**editor.widgets.focused**) does not seem to me to be bulletproof (for instance, what if some new editor version will not automatically set focus on pasted code/widget?).

Last, and I have to admit, I found this behaviour after filing this feature request, there seems to be different workflow for **widgets.focused.edit()** and for widget executed by its command, as there is no "**defaults**" processing and widget's template with some {variables} is not replaced with their defaults.


As I mentioned in my second comment, I found another approach/hack which directly executes editor's widget command which behaves exactly as it has been clicked by toolbar icon (with automatic dialog binding), but there is hacky **event.data.dataValue = ''** line... Anyway, using this "command" flow and thus simulating toolbat button click seems to me as better approach and I'm now using it even with manually resetting dataValue value. Can I have your comment on this as well please?

So to sum it up - I would like to have some direct clean support for this automatic dialog binding.

**Comment by Dumluregn** [comment]
> as afterPaste reacts on any paste event so there has to be some "if"

Well, the logic detecting if there is a widget in pasted / selected / dropped content has to be implemented somewhere anyway. We theoretically could add some sort of widget detection in the clipboard plugin to expose all widgets found in the dataTransfer to be also able to handle the case where need to do something with more than one widget, so I'll mark this issue as a valid feature request. However this looks like an edge case with the pretty good workaround already (stopping focusing the pasted content seems like a big change and we are unlikely to do this), so I have to tell right away the chances we'll find time for this aren't great. Anyway, thank you for creatively using CKEditor.
